torch/distributed/nn/api/remote_module.py

```diff
@@ -0,0 +1,245 @@
+#!/usr/bin/python3
+import uuid
+from typing import Any, Callable, Dict, Tuple
+
+import torch
+import torch.distributed.rpc as rpc
+from torch.distributed.nn.jit import instantiator
+
+
+_NON_SCRIPTABLE_REMOTE_MODULE_MODULE = instantiator.instantiate_non_scriptable_remote_module_template()
+_NON_SCRIPTABLE_REMOTE_MODULE_CLS = _NON_SCRIPTABLE_REMOTE_MODULE_MODULE._RemoteModule
+
+
+def _gen_global_unique_name():
+    return f"{uuid.uuid4().hex}"
+
+
+# RPC handler.
+def _instantiate_template(module_interface_cls):
+    instantiator.instantiate_scriptable_remote_module_template(
+        module_interface_cls
+    )
+
+
+def _scripted_module_creator_wrapper(module_creator, module_interface_cls, args, kwargs):
+    module = module_creator(*args, **kwargs)
+    script_module = torch.jit.script(module)
+    return rpc.RRef(script_module, module_interface_cls)
+
+
+def _RemoteModule(
+    to: str,
+    module_creator: Callable,
+    args: Tuple = None,
+    kwargs: Dict[str, Any] = None,
+    global_unique_name: str = None,
+    module_interface_cls: Any = None,
+):
+    """
+        A RemoteModule instance can only be created after RPC initialization.
+        It creates a user-specified module on a specified remote node.
+        It behaves like a regular ``nn.Module`` except that the ``forward`` method is
+        executed on the remote node.
+        It takes care of autograd recording to ensure the backward pass propogates
+        gradients back to the corresponding remote module.
+
+        The arguments of ``forward_async`` and ``forward`` are the same as
+        the ``forward`` method of the module returned by the ``module_creator``.
+
+        For example, if ``module_creator`` returns an instace of ``nn.Linear``,
+        that has ``forward`` method signature, ``def forward(input: Tensor) -> Tensor:``,
+        the generated ``RemoteModule`` will have 2 methods in signature of
+        ``def forward(input: Tensor) -> Tensor:`` and
+        ``def forward_async(input: Tensor) -> Future[Tensor]:``.
+
+    Arguments:
```

Comment:
I feel like a nicer API for RemoteModule could be

```
class RemoteModule(nn.Module):
	def __init__(to: str, local_module_cls, args, kwargs):
        self.to = to
        self.mod_rref = rpc.remote(to, local_module_cls, args, kwargs)

	def forward():
		# code for forward

	def forward_async():
		# code for async forward
```

For TorchScript support, we temporarily do

```
class _RemoteModule(nn.Module):
	def __init__(to: str, local_module_cls, args, kwargs, _module_interface_cls):
        self.to = to
        # call instantiator to instantiate the forward/forward_async methods from interface
		instantiated_forward_method = instantiator.instantiate_forward(interface_cls)
		instantiated_forward_async_method = instantiator.instatiate_forward_async(interface_cls)
		setattr(self, instantiated_forward_method)
		setattr(self, instantiated_forward_async_method)

Reply:
Not planning to do it. This way we will maintain two duplicate code. Instead, I only want to keep one ground truth in the template that can generate for both Python (without method signatures) and TorchScript(with method signatures).

Reply:
Updating
- Make both Python API and TorchScript API a class inherit from `nn.Module`.
- Simplify template to generate methods only, instead of generatin a whole class.
- Dynamically attach generated methods to the instance in `__init__`.